### PR TITLE
fix(CMSRepo): Support cache key generation for map params

### PR DESCRIPTION
**Pull request by kotva006**
#### Summary of changes
Parse the params of a map to a string for cache key generation
https://app.asana.com/0/555089885850811/1206990758684890

---

#### General checks
* [ ] **Are the changes organized into self-contained commits with descriptive and well-formatted commit messages?** This is a good practice that can facilitate easier reviews.
* [ ] **Testing.** Do the changes include relevant passing updates to tests? This includes updating screenshots. Preferably tests are run locally to verify that there are no test failures created by these changes, before opening a PR.
* [ ] **Tech debt.** Have you checked for tech debt you can address in the area you're working in? This can be a good time to address small issues, or create Asana tickets for larger issues.

#### New UI, or substantial UI changes
* [ ] **Cross-browser compatibility** is less of an issue now that we're no longer supporting IE, but changes still need to work as expected in Safari, Chrome, and Firefox.
* [ ] **Are interactive elements accessible?** This includes at minimum having [relevant keyboard interactions](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#keyboard-interaction) and [visible focus](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#visible-focus), but can also include verification with screen reader testing.
* [ ] **Other accessibility checks** such as sufficient [color constrast](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#color-contrast), or whether the layout holds up at 200% [zoom level](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#visible-focus).

#### New endpoints, or non-trivial changes to current endpoints
* [ ] **Have we load-tested** any new pages or internal API endpoints that will receive significant traffic? [See load testing docs](./../load_tests/README.md)
* [ ] **If this change involves routes,** does it work correctly with pertinent "unusual" routes such as the combined Green Line, Silver Line, Foxboro commuter rail, and single-direction bus routes like the 170?

### CODE REVIEW

**Review by amaisano — APPROVED** [review]

File: test/cms/repo_test.exs [hunk 1]
```diff
@@ -97,6 +97,16 @@ defmodule CMS.RepoTest do
                "cms.repo" <> String.replace(path, "/", "|") <> "?bam=bop&baz=qux"
     end
 
+    test "generates the correct key for a map" do
+      path = "/foo/bar"
+      params = %{"biz" => "bang", "data" => %{"some" => "map"}}
+
+      assert Repo.generate(nil, nil, [path, params]) ==
+               "cms.repo" <> String.replace(path, "/", "|") <> "?biz=bang&data=?some=map"
```
Comment (anthonyshull) [inline test/cms/repo_test.exs:105]: 💥 This won't be parseable by the CMS. `?biz=bang&data=?some=map` will just take `?some=map` as a string when you get the params.

File: lib/cms/repo.ex [hunk 2]
```diff
@@ -234,9 +234,11 @@ defmodule CMS.Repo do
   end
 
   defp params_to_string(params) when params == %{}, do: ""
+  defp params_to_string(nil), do: ""
```
Comment (anthonyshull) [inline lib/cms/repo.ex:237]: We should just make it `(_), do: ""` and put this after the string and map input types.

File: lib/cms/repo.ex [hunk 3]
```diff
@@ -234,9 +234,11 @@ defmodule CMS.Repo do
...
+  defp params_to_string(params) when is_binary(params), do: params
 
   defp params_to_string(params) when is_map(params) do
-    [head | tail] = Enum.map(params, fn {k, v} -> "#{k}=#{v}" end)
+    [head | tail] = Enum.map(params, fn {k, v} -> "#{k}=#{params_to_string(v)}" end)
```
Comment (anthonyshull) [inline lib/cms/repo.ex:241]: We should not handle nested maps here as they don't really make sense. I would investigate what is calling this with a nested map. It is probably wrong.

**Review by anthonyshull — CHANGES_REQUESTED** [review]

**Comment by anthonyshull** [comment]
Look at https://github.com/mbta/dotcom/blob/abee6a61e4057d1675f492fb6bb3bcc3fd6e7797/lib/cms/api/http_client.ex#L48 because that is where it is converting nested maps to a query string. We could either make that public and use it to generate the cache key. But, I am also suspicious about whether or not a nested map should be passed in as params.

File: test/cms/repo_test.exs (hunk 1 above)
Comment (anthonyshull) [reply test/cms/repo_test.exs:105]: This is the structure that is being passed in: `?schedule_finder[direction_id]=1&schedule_finder[origin]=3939`

Comment (anthonyshull) [reply test/cms/repo_test.exs:105]: Also, I can confirm that we pass those params on to the CMS and they are ignored.

Comment (amaisano) [reply test/cms/repo_test.exs:105]: > Also, I can confirm that we pass those params on to the CMS and they are ignored.

There is a whitelist filter for CMS params somewhere in the dotcom code. Not sure how it interacts with this stuff but you may want to track it down.

Comment (kotva006) [reply test/cms/repo_test.exs:105]: If the CMS ignores the params, then I am just going to reject the map

Comment (anthonyshull) [reply test/cms/repo_test.exs:105]: There might be cases where we don't ignore the params. That's why I'm thinking just make the params to string function that the CMS code is using public and then use that to make the key. That way they're guaranteed to be the same.

**Review by anthonyshull — CHANGES_REQUESTED** [review]
For the URL in the Sentry error, this code produces the following key:

"cms.repo|paragraphs|multi-column|bus-fares?schedule_finder=?direction_id=1&origin=3939"

It should match the actual URL params: `?schedule_finder[direction_id]=1&schedule_finder[origin]=3939`

**Comment by anthonyshull** [comment]
I would go through and debug the entire process with the URL in the Sentry error because it looks like the params are being used for the cache but they are getting removed when a request is sent to the CMS. Our goal is to make sure that the cache key makes sense with the request that is actually being sent.

**Comment by anthonyshull** [comment]
Let me know if you need more direction.

File: lib/cms/repo.ex [hunk 4]
```diff
@@ -234,15 +234,24 @@ defmodule CMS.Repo do
...
+        ["?#{head}", "#{Enum.join(tail, "&")}"]
+        |> Enum.reject(&(&1 == ""))
+        |> Enum.join("&")
 
-    ["?#{head}", "#{Enum.join(tail, "&")}"]
-    |> Enum.reject(&(&1 == ""))
-    |> Enum.join("&")
+      _ ->
+        ""
+    end
```
Comment (anthonyshull) [inline lib/cms/repo.ex:250]: You could do this using pattern matching to support a list.

**Review by anthonyshull — APPROVED** [review]

**Pull request merged**